db/dcrpg/queries.go

```diff
@@ -2475,6 +2476,81 @@ func retrieveTicketByOutputCount(db *sql.DB, dataType outputCountType) (*dbtypes
 	return items, nil
 }
 
+// retrieveChainWork assembles both block-by-block chainwork pts
+// and daily network hashrate points.
+func retrieveChainWork(db *sql.DB) (*dbtypes.ChartsData, *dbtypes.ChartsData, error) {
+	// Grab all chainwork points. Actually rows of (time, chainwork).
+	rows, err := db.Query(internal.SelectChainWork)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer closeRows(rows)
+
+	// Assemble chainwork and hashrate simultaneously.
+	// Chainwork is stored as a 32-byte hex string, so in order to
+	// do math, math/big types are used
+	workdata := new(dbtypes.ChartsData)
+	hashrates := new(dbtypes.ChartsData)
+	var blocktime uint64
+	var workhex string
+
+	// In order to store these large values as uint64, they are represented
+	// as exahash for work, and terahash/s for hashrate.
+	bigExa := big.NewInt(int64(1e18))
+	bigTera := big.NewInt(int64(1e12))
+	var dayStamp, firstDayStamp, lastStamp uint64
+	var firstDayWork, lastWork *big.Int
+	for rows.Next() {
+		// Get the chainwork
+		err = rows.Scan(&blocktime, &workhex)
+		if err != nil {
+			return nil, nil, err
+		}
+		bigwork := new(big.Int)
+		exawork := new(big.Int)
+		bigwork, ok := bigwork.SetString(workhex, 16)
+		if !ok {
+			log.Errorf("Failed to make big.Int from chainwork %s", workhex)
+			break
+		}
+		exawork.Set(bigwork)
+		exawork.Div(bigwork, bigExa)
+		if !exawork.IsUint64() {
+			log.Errorf("Failed to make uint64 from chainwork %s", workhex)
+			break
+		}
+		workdata.ChainWork = append(workdata.ChainWork, exawork.Uint64())
+		workdata.Time = append(workdata.Time, blocktime)
+
+		// Hashrates are binned into days
+		if blocktime > dayStamp {
+			if lastStamp > 0 {
```

Comment:
I suspect others might have opinions on this decision. Single-block hashrate data is worthless, and even up to many tens of blocks, it's pretty noisy. I chose to bin hashrate into days here, but binning into some number of blocks might also be appropriate. For reference, by default, the RPC call `getnetworkhashps` returns the most recent 120-block average.